Build: configure: auto-detect PCMK__API_VERSION

kgaillot opened the pull request.

kgaillot commented:
@clumens, what do you think of this? We'd just have to remember to re-run configure after changing API schemas

wenningerk commented:
One thing coming to my mind in that context:
With c-files the Make-file contains a list of files. So with a dangling unmanged c-files the build-output wouldn't change in any way.
With the schema-stuff we collect whatever we find in the directory - under source-control or not.
With your suggestion building of the schema and deriving of the version become more consistent in a way.
If we have a list in the Makefile we could define a rule that creates a header from these files that could be included.
To make it more safe and to point people to a change in behavior we could fail the build if there are non-listed rng-files.

kgaillot commented:
> One thing coming to my mind in that context: With c-files the Make-file contains a list of files. So with a dangling unmanged c-files the build-output wouldn't change in any way. With the schema-stuff we collect whatever we find in the directory - under source-control or not. With your suggestion building of the schema and deriving of the version become more consistent in a way. If we have a list in the Makefile we could define a rule that creates a header from these files that could be included. To make it more safe and to point people to a change in behavior we could fail the build if there are non-listed rng-files.

Good point, but we already have that problem with xml/Makefile.am -- it does an ls as well to determine the files. Having to explicitly list every RNG in the Makefile would be painful, there's nearly 100 of them currently -- the current approach would be preferable I think. It would be nice if we could use git to list the managed files, but we can't require git for a basic build. I suppose we could use git with ls as a fallback, but that makefile is already difficult to follow.

wenningerk commented:
> Good point, but we already have that problem with xml/Makefile.am -- it does an ls as well to determine the files. Having to explicitly list every RNG in the Makefile would be painful, there's nearly 100 of them currently -- the current approach would be preferable I think. It would be nice if we could use git to list the managed files, but we can't require git for a basic build. I suppose we could use git with ls as a fallback, but that makefile is already difficult to follow.

Haven't said this behavior was new. Even said we would do it more consistently now - both for collecting the xml & deriving the version ;-) The topic just reminded me to address this basic inconsistency in the build process.
Anyway what we could do is provide a shell-script that updates the filelists from git - maybe show a list of differences first and just do the update with an extra switch.

kgaillot commented:
> Anyway what we could do is provide a shell-script that updates the filelists from git - maybe show a list of differences first and just do the update with an extra switch.

Yeah, that was kind of my long-term goal, to have a script that would copy the current highest schema to the next higher version and get everything set up so we could just edit and commit. I think this would be a decent improvement in the interim, since most users will not jump back and forth in checkouts like we do.

nrwahl2 commented:
> We'd just have to remember to re-run configure after changing API schemas

I wonder if there's any way to tell the tool chain to do that on the next `make` command after the schemas change... `configure` sometimes re-runs automatically on `make` (usually after checking out something very different) but I don't know what triggers it.

kgaillot commented:
> > We'd just have to remember to re-run configure after changing API schemas
> 
> I wonder if there's any way to tell the tool chain to do that on the next `make` command after the schemas change... `configure` sometimes re-runs automatically on `make` (usually after checking out something very different) but I don't know what triggers it.

Turns out yes, there is a CONFIG_STATUS_DEPENDENCIES variable for that purpose.

Rebased on main and updated commit to do that. It reconfigures after any change to any rng in xml/api, not just new files, but since we don't change older rngs that shouldn't be much of an annoyance.

nrwahl2 commented:
Excellent!

kgaillot commented:
OK, I added a subshell to use git if available and fall back to ls. That should take care of all concerns

nrwahl2 commented on configure.ac:
```diff
@@ -319,6 +319,18 @@ VERSION_ARG(VERSION_NUMBER)
+# Detect highest API schema version (use git if available to list managed RNGs,
+# in case there are leftover schema files from an earlier build of a different
+# version, otherwise check all RNGs)
+API_VERSION=$({ git ls-files xml/api/*.rng 2>/dev/null || ls -1 xml/api/*.rng ; } dnl
```
Nice!

It'd be nice to see something like this in the schema generation code too at some point, if we're not already doing it. I occasionally fight weird issues and then realize it built the schemas based on a leftover newer version.

wenningerk commented:
> OK, I added a subshell to use git if available and fall back to ls. That should take care of all concerns

Apart from that you can't test with a new file without bringing it to git and that nothing will point you
to the fact that you might have forgotten to do so ;-)

clumens commented:
> > We'd just have to remember to re-run configure after changing API schemas
> 
> I wonder if there's any way to tell the tool chain to do that on the next `make` command after the schemas change... `configure` sometimes re-runs automatically on `make` (usually after checking out something very different) but I don't know what triggers it.

That's because when you type "make", it first looks for a GNUmakefile before anything else.  We have one of those, so it goes to that file and looks for the default target which is usually "all" but we've changed it to "default".  That target depends on the "init" target, and that target re-runs autogen.sh and ./configure.

clumens commented:
This looks good to me.

nrwahl2 commented:
> > OK, I added a subshell to use git if available and fall back to ls. That should take care of all concerns
> 
> Apart from that you can't test with a new file without bringing it to git and that nothing will point you to the fact that you might have forgotten to do so ;-)

That is a good point...

nrwahl2 commented:
> That's because when you type "make", it first looks for a GNUmakefile before anything else. We have one of those, so it goes to that file and looks for the default target which is usually "all" but we've changed it to "default". That target depends on the "init" target, and that target re-runs autogen.sh and ./configure.

Ah, good to know :) Any idea why it usually doesn't re-run (or conversely, what triggers it occasionally)? I feel like I've seen it happen even when `configure`, `libtldl`, and `Makefile` already exist, but I could be wrong about that.

kgaillot commented:
> > Apart from that you can't test with a new file without bringing it to git and that nothing will point you to the fact that you might have forgotten to do so ;-)
> 
> That is a good point

It being ignored will help point you ;)

We always make changes in a two-step process, first committing an exact copy of the previous schema, then making changes, so it shouldn't be a problem when that's followed

kgaillot commented:
> > That's because when you type "make", it first looks for a GNUmakefile before anything else. We have one of those, so it goes to that file and looks for the default target which is usually "all" but we've changed it to "default". That target depends on the "init" target, and that target re-runs autogen.sh and ./configure.
> 
> Ah, good to know :) Any idea why it usually doesn't re-run (or conversely, what triggers it occasionally)? I feel like I've seen it happen even when `configure`, `libtldl`, and `Makefile` already exist, but I could be wrong about that.

I think that part is actually a built-in rule of autoconf, it'll re-run configure if any .in changed since the last run
